### PR TITLE
Improve polygon extraction

### PR DESCRIPTION
**Pull request by Evarin**
Hello,

I am currently working for the project FONDUe at Université de Genève, on their eScriptorium instance.

While investigating on their performance issues, it appeared that a lot of time is spent warping the extracted lines. More specifically, scikit-image functions `draw.polygon` and `warp` are really inefficient, and the kraken functions could be much faster if they used Pillow `Image.transform` -- or at least scipy.ndimage `affine_transform`.

While working on those improvements, I also noticed that the current kraken `extract_polygon` function caused strange warping artifacts on complex baselines, due to the use of skimage `PiecewiseAffineTransform`. In order to improve the quality of the warping, and make it compatible with Pillow `transform` syntax, I wrote a smoother warping function (`_bevelled_warping_envelope`) that deforms much less the characters of the target line.

Here is an example of the results (older version above, new proposition below, the plot shows the computed envelope transformation):
kraken_faster

You can see more of them on [this colab](https://colab.research.google.com/drive/1eJHqgGoOsCwfuuz4ubG2xxanNCoP8SMT?usp=sharing): my modified functions run 30 to 60 times faster that the older ones, while also enabling antialiasing (which was disabled in kraken 4.3.10).

I integrated them in an older version of kraken, and it works properly (and greatly improve transcription performance). But I see you are currently working on changing the input formats, and the tests are currently broken, so I could not check that everything is working well with this PR. I still propose it now for the record, and would be very happy to work on it further when tests are ready :)

### CODE REVIEW

**Comment by mittagessen** [comment]
Thanks, this is great! The anti-aliasing was disabled because of B/W lines which obviously aren't anymore with it. Disabling it didn't seem to have any accuracy impact on grayscale models so just turning it off was easiest.

The important question is if it impacts the accuracy of existing models. Have you verified that? In general recognition models are quite sensitive to changes in polygonization/preprocessing so it is possible they might not be "compatible" with your new, faster alternative. I'd still merge it but we'd need to add a flag in the model metadata so the recognizer knows which extractor to use.

**Comment by Evarin** [comment]
Ok, thanks, I'll run it with BiblIA! (but it's a pretty big one, it will take a bit of time)

So far, with CREMMA and ManuMcFrench, I get 91.39% with the old method, and 90.35% with the new one (knowing that it uses the "slow path" in 25% of the cases, and the main recognition errors in both cases seem to be a format issue between dataset and model: `786 { é } - { e }` / `702  {  } - { COMBINING ACUTE ACCENT }`)

**Comment by mittagessen** [comment]
On 23/12/05 01:16AM, Robin Champenois wrote:
> Ok, thanks, I'll run it with BiblIA! (but it's a pretty big one, it
> will take a bit of time)

A subset is fine. The model has been trained on most of that data so
running everything will just get you the accuracy on the training set
anyway.

> So far, with CREMMA and ManuMcFrench, I get 91.39% with the old
> method, and 90.35% with the new one (knowing that it uses the "slow
> path" in 25% of the cases, and the main recognition errors in both
> cases seem to be a format issue between dataset and model: `786 { é }
> - { e }` / `702  {  } - { COMBINING ACUTE ACCENT }`)

Thanks, I'll run some tests myself. 1 percentage point drop isn't good
and would most likely require the aforementioned switch but I'll fiddle
around with the smoothing/anti-aliasing to see if that changes things
and also validate against training.
Just because it looks better to a human might not necessarily mean that
it improves recognition quality (for models trained with your
implementation that means).

BTW, I've fixed inference on main, so if you'd prefer working on that it
would be possible.

**Comment by Evarin** [comment]
On a (1100 lines) BiblIA subset, I get 98.36% with the old method, 97.91% with the new one (97.89% with the new one, disabling antialias). So, a 0.35% difference. Most of the processing was done on the multi-line path.

**Comment by mittagessen** [comment]
Sorry from my side on not following up on this either. I ran a couple of experiments myself in December to see which component actually causes the loss of accuracy but couldn't identify one particular thing. 

My plan was to implement the second option. On good models the 0.5 percentage points decrease in accuracy is an increase in overall errors between 33% and 50% which would be fairly unacceptable, especially as most people do not really care if it takes 0.5 or 1 second per line (I know large digitization projects are different). Add a flag to the model (and binary dataset) metadata when it was trained with the new line normalizer, set it as a default on the training side, and fall back to the old implementation on inference when the flag is absent. Plus printing a big fat warning on the command line that the user might consider retraining with the new method because it is so much faster.

**Comment by Evarin** [comment]
Ok, let's do that then. How do you want to proceed? I can update my PR to be up to date with `main` and implement this hybrid behavior, is that fine for you?

**Comment by mittagessen** [comment]
Yes, if it isn't too much trouble for you I'd appreciate you updating the pull request for current `main`. If not I can do it by mid-March.

It is actually the last thing I wanted to get into the main branch before tagging a new stable release.

**Comment by Evarin** [comment]
Ok I'll work on it then!

**Comment by Evarin** [comment]
Oh, right, I forgot those. 

I've added support, but as they are precompiled it's less convenient to track their status :/. For now I've added the flag to the Arrow IPC file metadata, and only throw a warning if we try to train/test expecting new polygons but use an old-fashioned dataset. See the tests, that checks for the presence of the warning.

It would be more convenient if we could toggle it based on the actual dataset info, but it seems quite complicated to do so, and might be unreliable. Or, we could make the warnings an error. Or keep it like that. What do you think?

**Comment by mittagessen** [comment]
I would have put the test in the trainer and just exposed the set of legacy polygon flags from the arrow file(s) metadata in the dataset. It reduces the need to hand over more arguments.

**Comment by Evarin** [comment]
Thanks for your comment, I had not seen the edited version of it (I just read my mail...), so I started implementing something slightly too complicated :/

My motivation for letting the parameter in the Dataset class is that people may want to evaluate the difference in accuracy between the old method and the new one. So maybe : keep the optional flag in test and in Dataset class, but remove it elsewhere, and force-upgrade when training (except in the very special case of old arrow datasets), with warnings when arrow dataset are old or incongruent?

**Comment by Evarin** [comment]
Okay, I can make those changes, or wait for you to look through the code and comment precisely the changes you'd like to see -- and avoid other misunderstandings ^^". What do you prefer?

**Comment by mittagessen** [comment]
On 24/03/06 02:48AM, Robin Champenois wrote:
> Okay, I can make those changes, or wait for you to look through the code and comment precisely the changes you'd like to see -- and avoid other misunderstandings ^^". What do you prefer?

I think I'd prefer looking through the code tomorrow. If there are just
a few small changes I might just merge as is and do them myself as
people are complaining that I should tag a new release ;)

**Comment by mittagessen** [comment]
Sorry for the delay. I've merged everything and it seem to work as intended. I'll write a release note and then tag a new release.

**Comment by anutkk** [comment]
This new mechanism indeed speeds up things about 10x, but in my case (BiblIA on Geneve 146) it changes significantly the output.
For example, the line:
image

which is transcribed using the legacy method as:

> יש נהכי השהא שנ ושב יעקב ושרר ושאו ו הא והרשעי

Is transcribed as following by the new method:

> 'יש נהכי השהא שנ ושב יעקב ושר ושאנ : ו הא והרשעי

Admittedly not a good result to begin with, I just want to illustrate it does affect the model.

I have coded a much lighter optimization (basically just changing `np.array` to`np.asarray`; `skimage.draw.polygon` to `PIL.ImageDraw.Draw`; and using a [faster, vectorized version of `skimage.transform.PiecewiseAffineTransform`](https://github.com/scikit-image/scikit-image/issues/6864). It speeds up by "only" 3x, but it does not change the output at all.

If we prove that it is really equivalent (`PIL.ImageDraw.Draw` is different by only a few pixels, and the new `PiecewiseAffineTransform` has been shown to change the output by [less than 1e-15](https://github.com/scikit-image/scikit-image/pull/6963#issuecomment-1872077806) - maybe this could be a transparent improvement to the legacy extraction?

@Evarin  could you share the code you used to measure the error rate on the BiblIA dataset?

**Comment by mittagessen** [comment]
> If we prove that it is really equivalent (PIL.ImageDraw.Draw is different by only a few pixels, and the new PiecewiseAffineTransform has been shown to change the output by https://github.com/scikit-image/scikit-image/pull/6963#issuecomment-1872077806 - maybe this could be a transparent improvement to the legacy extraction?

Yes, if there's no change in recognition accuracy I'd say we can modify the legacy code path to at least get a partial speedup.

**Comment by Evarin** [comment]
I tested the old and new code from my branch [f-improve](https://github.com/Evarin/kraken/tree/f-improve), and using `ketos test`
Basically, you can toggle the different modifications I made using https://github.com/Evarin/kraken/blob/dbeec21c8f8b91d6a80d7f717e310a8817d23ba6/kraken/lib/segmentation.py#L49 (with `warp=legacy|fast_legacy|new`, and the others `legacy|new`)

The "fast_legacy" is already an attempt to improve the `PiecewiseAffineTransform`, but using PIL, and it was already leading to significant changes, so we did not keep it. 

(Also, in my evaluations, I don't know what was in the training set of the model I use, I can't measure how much of it is from the model being overfit to the exact pixels, so I trust you for saying it's a problem or not...)

**Comment by anutkk** [comment]
> > If we prove that it is really equivalent (PIL.ImageDraw.Draw is different by only a few pixels, and the new PiecewiseAffineTransform has been shown to change the output by [scikit-image/scikit-image#6963 (comment)](https://github.com/scikit-image/scikit-image/pull/6963#issuecomment-1872077806) - maybe this could be a transparent improvement to the legacy extraction?
> 
> Yes, if there's no change in recognition accuracy I'd say we can modify the legacy code path to at least get a partial speedup.

@mittagessen I tested on the Italian subset (3748 lines). Results:

| Method| Accuracy| Runtime (minutes:seconds)
| ----------- | ----------- | ----------- |
| Current legacy extraction| 97.29% | 56:27 |
| Slightly optimized legacy extraction| 97.33% | 29:26 |

The speedup is more pronounced for straight baselines (closer to 4x). 

The difference in accuracy is only 0.04% (and increased for whatever reason). Close enough?

**Comment by mittagessen** [comment]
0.04 is below 5% change for a good model so I'd deem that acceptable given the speedup. I'm a bit busy the next couple of days. If you could create a pull request I'll merge it but otherwise I won't get around to it until the end of the month.

**Comment by anutkk** [comment]
@mittagessen Sure thing, no problem. On the main branch or the v5.0 branch?

**Comment by mittagessen** [comment]
On 24/04/04 12:46PM, anutkk wrote:
> @mittagessen Sure thing, no problem. On the main branch or the v5.0 branch?

main is better. I'm squishing small bugs I've missed in the 5.0 release
but otherwise it will be stable.

**Comment by anutkk** [comment]
> On 24/04/04 12:46PM, anutkk wrote: @mittagessen Sure thing, no problem. On the main branch or the v5.0 branch?
> main is better. I'm squishing small bugs I've missed in the 5.0 release but otherwise it will be stable.

@mittagessen Done: https://github.com/mittagessen/kraken/pull/586
